src/UartSerial.cpp

```diff
@@ -104,22 +102,30 @@ void UartSerial::begin(uint32_t speed_, size_t config_)
 
 void UartSerial::end()
 {
-  if(_fd != -1){
-    flush();
-    // Restore the original settings
-    tcsetattr(_fd, TCSANOW, &_tio_original_config);
-    close(_fd);
-    _fd = -1;
-  }
+  if(_fd == -1) return;
+  
+  flush();
+
+  // Shut down the poll() thread
+  _polling = false;
+  _pollThread.join();
+
+  // Restore the original settings
+  tcsetattr(_fd, TCSANOW, &_tio_original_config);
+  close(_fd);
+  _fd = -1;
 }
 
 void UartSerial::flush()
 {
-  tcdrain(_fd);
+  if(_fd != -1)
+    tcdrain(_fd);
 }
 
 int UartSerial::read()
 {
+  if(_fd == -1) return -1;
+
   int buffer[1];
```

Comment:
I figured out why I think this is weird. I can see you chose `int` as the buffer type, because it is also the return value. However, because you are using the array syntax instead of using `&buffer` where `buffer` has the storage of an `int`, then if you were to read two bytes instead of 1, then you would have both bytes stored at the first index. Also `int` is a variable size type, so it is impossible to tell how many bytes would actually go into a single `int`.

I think you should change the type of `buffer` to `uint8_t`, or remove the array notation, or both.